### PR TITLE
REFACTOR: New assign priority function

### LINKED ISSUE #4645: Bug located in assign_priority
### Before submitting the issue

- [X] I have searched among the existing issues
- [X] I am using a Python virtual environment

### Description of the bug

Assign priority method does not work as expected due to a limitation of native APIs.

The function only works when applying one level of higher priority (level 2) as it destroys any previously set priority.

### Steps To Reproduce

-

### Which Operating System are you using?

Windows

### Which Python version are you using?

3.7

### Installed packages

-

### PR DESCRIPTION
**Pull request by lorenzovecchietti**
Close  #4645

### CODE REVIEW

File: pyaedt/modules/MeshIcepak.py [hunk 1]
```diff
@@ -1280,10 +1283,94 @@ def automatic_mesh_3D(self, accuracy, enable_stair_step=True):
...
+    def assign_priorities(self, assignment):
+        """Set objects priorities.
+
+        Parameters
+        ----------
+        assignment : List[List[Union[Object3d, UserDefinedComponent, str]
+            List of lists of objects. Each list corresponds to one priority level from low to high.
+            This means that the first list will have the lowest priority while the last list will
+            have the higher priority. Objects not explicitly passed in the lists will be assigned
+            to a priority level lower than the objects in the first list.
```
Comment (PipKat) [inline pyaedt/modules/MeshIcepak.py:1296]:
```suggestion
            This means that the first list has the lowest priority while the last list
            has the highest priority. Objects not explicitly passed in the lists are assigned
            to a priority level lower than the objects in the first list.
```
Avoid future tense.

File: pyaedt/modules/MeshIcepak.py [hunk 2]
```diff
@@ -1280,10 +1283,94 @@ def automatic_mesh_3D(self, accuracy, enable_stair_step=True):
...
+        _dict2arg(props, args)
+        self.modeler.oeditor.UpdatePriorityList(args[0])
+        return True
+
     @pyaedt_function_handler(obj_list="assignment", comp_name="component")
     def add_priority(self, entity_type, assignment=None, component=None, priority=3):
         """Add priority to objects.
 
+        .. deprecated:: 0.9.1
+        Use :func:`assign_priorities` method instead.
```
Comment (PipKat) [inline pyaedt/modules/MeshIcepak.py:1372]:
```suggestion
        Use :func:`assign_priorities` function instead.
```
Assuming this is a function based on the role used.

File: pyaedt/modules/MeshIcepak.py [hunk 3]
```diff
@@ -1315,6 +1402,7 @@ def add_priority(self, entity_type, assignment=None, component=None, priority=3)
         >>> app.mesh.add_priority(entity_type=1,assignment=app.modeler.object_names,priority=3)
         >>> app.mesh.add_priority(entity_type=2,component=app.modeler.user_defined_component_names[0],priority=2)
         """
+        warnings.warn("Use :func:`assign_priorities` method instead.", DeprecationWarning)
```
Comment (PipKat) [inline pyaedt/modules/MeshIcepak.py:1405]:
```suggestion
        warnings.warn("Use :func:`assign_priorities` function instead.", DeprecationWarning)
```

File: pyaedt/modules/MeshIcepak.py [hunk 4]
```diff
@@ -1280,10 +1283,94 @@ def automatic_mesh_3D(self, accuracy, enable_stair_step=True):
...
+
+        >>> oEditor.UpdatePriorityList
+
+        Examples
+        --------
+
+        >>> ipk.mesh.assign_priorities([["Box1", "Rectangle1"], ["Box2", "Fan1_1"], ["Heatsink1_1"]])
+        """
+        if not isinstance(assignment[0], list):
+            raise AttributeError("``assignment`` input must be a list of lists.")
```
Comment (SMoraisAnsys) [inline pyaedt/modules/MeshIcepak.py:1314]:
```suggestion
        if not assignment or not isinstance(assignment, list) or not isinstance(assignment[0], list):
            raise AttributeError("``assignment`` input must be a list of lists.")
```
This should cover a wider rander of types for `assignment`. Currently, a `TypeError` will be raised if `assignment` is not subscriptable. If that's the behavior you wanted, please discard this comment.

File: pyaedt/modules/MeshIcepak.py [hunk 5]
```diff
@@ -1280,10 +1283,94 @@ def automatic_mesh_3D(self, accuracy, enable_stair_step=True):
...
+                o
+                for o in objects
+                if (isinstance(o, Object3d) and o.is3d)
+                or (isinstance(o, UserDefinedComponent) and any(p.is3d for p in o.parts.values()))
+            ]
+            obj_2d = [
+                o
+                for o in objects
+                if (isinstance(o, Object3d) and not o.is3d)
+                or (isinstance(o, UserDefinedComponent) and any(not p.is3d for p in o.parts.values()))
```
Comment (SMoraisAnsys) [inline pyaedt/modules/MeshIcepak.py:1333]:
```suggestion
                or (isinstance(o, UserDefinedComponent) and all(not p.is3d for p in o.parts.values()))
```
Shouldn't that be checking that all parts are non 3D ?

**Review by SMoraisAnsys — COMMENTED** [review]
I think there might be a typo, feel free to resolve the conversation if I'm wrong.

**Pull request merged**